app/src/main/java/com/example/habitsmasher/MapDialog.java

```diff
@@ -0,0 +1,210 @@
+     * @param callback
+     */
+    private void getCurrentLocation(FirestoreCallback callback) {
+        _fusedLocationClient.getCurrentLocation(LocationRequest.PRIORITY_HIGH_ACCURACY,
+                                  null)
+                             .addOnSuccessListener(location -> {
+                                 if (location != null) {
+                                     _selectedLocation = location.getLatitude() +
+                                                         " " + location.getLongitude();
+                                     callback.onCallback(false);
+                                 }
+                                 else {
+                                     callback.onCallback(true);
+                                 }
+                             });
+    }
+
+
+    @Override
+    public void onSaveInstanceState(Bundle outState) {
+        super.onSaveInstanceState(outState);
+
+        Bundle mapViewBundle = outState.getBundle(MAPVIEW_BUNDLE_KEY);
+        if (mapViewBundle == null) {
+            mapViewBundle = new Bundle();
+            outState.putBundle(MAPVIEW_BUNDLE_KEY, mapViewBundle);
+        }
+
+        _mapView.onSaveInstanceState(mapViewBundle);
+    }
+
+    @Override
+    public void onAttach(@NonNull Context context) {
+        super.onAttach(context);
+        try {
+            // Get habit event fragment for later use
+            _habitEventDialog = (HabitEventDialog) getTargetFragment();
+        }
+        catch (ClassCastException e) {
+            Log.e(TAG, "Exception" + e.getMessage());
+        }
+        _fusedLocationClient = LocationServices.getFusedLocationProviderClient(context);
+    }
+
+    @Override
+    public void onResume() {
+        super.onResume();
+        _mapView.onResume();
+    }
+
+    @Override
+    public void onStart() {
+        super.onStart();
+        _mapView.onStart();
+    }
+
+    @Override
+    public void onStop() {
+        super.onStop();
+        _mapView.onStop();
+    }
+    
+    @Override
+    public void onPause() {
+        super.onPause();
+        _mapView.onPause();
+    }
+
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        _mapView.onDestroy();
+    }
+
+    @Override
+    public void onLowMemory() {
+        super.onLowMemory();
+        _mapView.onLowMemory();
+    }
+
+    @Override
+    public void onMapReady(@NonNull GoogleMap googleMap) {
+        if (_selectedLocation.equals("")) {
+            getCurrentLocation(isLocationNull -> {
+                if (isLocationNull) {
+                    // location is recorded as null for some reason
+                    _habitEventDialog.displayErrorMessage(HabitEventDialog.LOCATION_NOT_RECORDED);
+                    getDialog().dismiss();
+                }
+                else {
+                   setUpMap(googleMap);
+                }
+            });
+        }
+        else {
+            setUpMap(googleMap);
+        }
```

Comment:
This essentially checks whether the habit event has specified a location before, and if it has not, queries the current location of the user